### ISSUE TITLE
[FR] Improve support for simultaneous Conda and Docker environment

### ISSUE THREAD

**Issue by Zethson** [mlflow/mlflow#2850]
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [x] No. I cannot contribute this feature at this time.

## Proposal Summary

Would be very useful and for my use case mandatory, if both, a Conda environment and a Docker environment could be present in the project directory. The user can then pick which of the two environments to run mlflow with e.g. with flags: -with-conda or -with-docker

## Motivation
- What is the use case for this feature?
- Why is this use case valuable to support for MLflow users in general?
- Why is this use case valuable to support for your project(s) or organization?
- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)

It would allow us to build Docker containers based on the same present conda environment file. This would reduce duplicated implementation efforts. Furthermore, the user would gain the choice of which environment to run the mlflow run with. 

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

## Details

Nextflow uses either flags for this (-with-conda or -with-docker) or predefined profiles for conda/docker/singularity environments such as -profile docker or -profile conda (they define run time parameters, which is another feature I would love to already have as configuration in the MLproject file, but that's another story)

**Comment by smurching** [comment]
@Zethson thanks for filing this, just to make sure we understand your use case / goals, is the idea that you'd like to define a project with some base docker environment, and then iterate on / make changes to the Python environment within that image without needing to rebuild/push a new docker image?

**Comment by Zethson** [comment]
Again, thank you for your fast feedback @smurching !

> @Zethson thanks for filing this, just to make sure we understand your use case / goals, is the idea that you'd like to define a project with some base docker environment, and then iterate on / make changes to the Python environment within that image without needing to rebuild/push a new docker image?

I don't think that we have the same idea.
I'll try to explain it again: 
## issue
Please very briefly glance over [this example](https://github.com/Zethson/mlflow_custom_ms_example). This is basically the mlflow multistep example, but with a Docker container and a conda environment.
Now, please examine the conda.yaml file. This is 100% the very same file that your multistep example WITHOUT DOCKER has. So no changes here.
But now we have the Dockerfile, which simply takes the conda.yaml file and builds the Conda environment to allow the multistep example to also be run from Docker! (Given https://github.com/mlflow/mlflow/issues/2851 get solved).
Hence, the mlflow project can be run via conda (as before in your example), but also via  Docker (since it simply builds the conda environment).
**The issue:** when a Docker image **AND** a conda environment is specified in the MLProject file, the error: `2020/05/19 10:17:08 ERROR mlflow.cli: === Project cannot contain both a docker and conda environment. ===` occurs.

## solution
mlflow should 100% allow a conda env **AND** a Docker image to be specified in the MLProject file.
By default it should always try to run the most easily accessible solution (which would be conda I guess) OR run the --no-conda version, which tries to fetch all dependencies from the system.

Furthermore, new flags should be introduced:
1. with-conda, which would run the current mlflow project with the specified conda env file
2. with-docker, which would run the current mlflow project with the specified docker container
3. with-singularity, which would take the current Docker container and run it via Singularity (NOTE: This would require mlflow to support Singularity, which it does no yet do, but I wanted to mention it here)

## Your question

> And if not would you be able to clarify the end user workflow you're envisioning once this feature is implemented (e.g. from writing the project, to executing the project, to ultimately sharing the project with collaborators to execute)?

The end user would not notice this change, except that he would now need to use the flag -with-conda or with-docker to run a specific environment. Heck, if just a single environment is defined in the MLProject file (like the current situation only allows), then no changes would be required for the user and he wouldn't even notice this, since we could just run the specified environment.
Sharing the project would be like now, I do not see anything that needs to be changed.
The user would still run `mlflow run https://github.com/blub `, but now add the flags -with-conda if he wants to run it with the conda environment or add the flag -with-docker if he wants to run it within a Docker environment (if present!).
About writing the project: Ideally, the mlflow community would adopt a structure like the Nextflow community, which supports the hierarchical approach of conda envs, which are loaded within an also present Docker container to facilitate both options. Since Nextflow also supports Singularity (which can run Docker containers), only the two files are required to support Conda, Docker and Singularity. for the end user. I am working on a tool to facilitate this hierarchical structure for the mlflow community (with GPU support etc) and have already been in contact with Paul Ogilvie who supports my idea. Hence, I hope that we can solve my issues quickly.

## Why it matters
Supporting the definition of several different environments such as Conda, Docker (& more) allows the user to run the project in his/her preferred environment. It just gives flexibility. Also, Docker and Singularity to of course strongly improve reproducibility, Conda just doesn't cut it.
The most important reasons are that many high performance environments only allow specific environments to be run. e.g. I am working in a bioinformatics setting on clusters, which also regularly host patient data. We do not allow our Nextflow pipelines to be run with Conda, nor Docker but only Singularity for security reasons.
Now, other big institutions only support Docker, since their admins do not want to deal with Singularity, hence they run the pipelines with Docker.

To summarize: The end user would only have to edit the conda.env file, since it would be loaded from the Docker container, which would in turn (in a hopeful close future) be loadable from Singularity.
As a result, all 3 environments would be support. This enables flexibility and is actually 100% required for big, safety concerned institutions.
I am working

**Comment by rdwrt** [comment]
Just a question: is this proposing the use case to deploy Singularity as a backend to MLFlow? 

I am currently interested in updating MLFlow status of processes running in Singularity, but that is something different, right?

**Comment by dbczumar** [comment]
Hi folks! @smurching @Zethson I think that support conda and Docker environments in MLprojects file would be a great enhancement. I agree that flags of the form `--with-docker` and `--with-conda` or a `--env` flag that can be used to specify the environment would be very useful. We could also deprecate `--no-conda` and replace this with `--no-env`. It would be great to get a concrete design for this feature (proposed APIs, proposed behaviors when one environment is specified + when multiple environments are specified).

I've added a `help wanted` label to request a design from the community.
